File: persistence/scaladsl/src/test/scala/com/lightbend/lagom/scaladsl/persistence/TestEntity.scala
/*
 * Copyright (C) 2016-2017 Lightbend Inc. <https://www.lightbend.com>
 */
package com.lightbend.lagom.scaladsl.persistence

import akka.actor.ActorRef
import akka.actor.ActorSystem
import akka.actor.Address
import akka.cluster.Cluster
import com.lightbend.lagom.scaladsl.persistence.PersistentEntity.ReplyType
import com.lightbend.lagom.scaladsl.persistence.testkit.SimulatedNullpointerException
import com.lightbend.lagom.scaladsl.playjson.{ JsonSerializerRegistry, JsonSerializer }

import scala.collection.immutable

object TestEntity {

  object SharedFormats {
    import play.api.libs.json._

    implicit val modeFormat = Format[Mode](
      Reads[Mode] {
        case JsString("append")  => JsSuccess(Mode.Append)
        case JsString("prepend") => JsSuccess(Mode.Prepend)
        case js                  => JsError(s"unknown mode js: $js")
      }, Writes[Mode] {
        case Mode.Append  => JsString("append")
        case Mode.Prepend => JsString("prepend")
      }
    )
  }

  object Cmd {
    import play.api.libs.json._
    import JsonSerializer.emptySingletonFormat
    import SharedFormats._

    val serializers = Vector(
      JsonSerializer(Json.format[Add]),
      JsonSerializer(Json.format[ChangeMode]),
      JsonSerializer(emptySingletonFormat(Get)),
      JsonSerializer(emptySingletonFormat(UndefinedCmd)),
      JsonSerializer(emptySingletonFormat(GetAddress))
    )

  }

  sealed trait Cmd

  case object Get extends Cmd with ReplyType[State]

  final case class Add(element: String, times: Int = 1) extends Cmd with ReplyType[Evt]

  sealed trait Mode
  object Mode {
    case object Prepend extends Mode
    case object Append extends Mode
  }

  final case class ChangeMode(mode: Mode) extends Cmd with ReplyType[Evt]

  case object UndefinedCmd extends Cmd with ReplyType[Evt]

  case object GetAddress extends Cmd with ReplyType[Address]

  object Evt {
    val NumShards = 4
    // second param is optional, defaults to the class name
    val aggregateEventShards = AggregateEventTag.sharded[Evt](NumShards)

    import play.api.libs.json._
    import JsonSerializer.emptySingletonFormat
    import SharedFormats._

    val serializers = Vector(
      // events
      JsonSerializer(Json.format[Appended]),
      JsonSerializer(Json.format[Prepended]),
      JsonSerializer(emptySingletonFormat(InPrependMode)),
      JsonSerializer(emptySingletonFormat(InAppendMode))
    )
  }

  sealed trait Evt extends AggregateEvent[Evt] {
    override def aggregateTag: AggregateEventShards[Evt] = Evt.aggregateEventShards
  }

  final case class Appended(element: String) extends Evt

  final case class Prepended(element: String) extends Evt

  case object InPrependMode extends Evt

  case object InAppendMode extends Evt

  object State {
    val empty: State = State(Mode.Append, Nil)

    import play.api.libs.json._
    import JsonSerializer.emptySingletonFormat
    import SharedFormats._
    val serializers = Vector(
      JsonSerializer(Json.format[State])
    )
  }

  final case class State(mode: Mode, elements: List[String]) {
    def add(elem: String): State = mode match {
      case Mode.Prepend => new State(mode, elem +: elements)
      case Mode.Append  => new State(mode, elements :+ elem)
    }
  }

  // TestProbe message
  final case class AfterRecovery(state: State)
}

object TestEntitySerializerRegistry extends JsonSerializerRegistry {
  import TestEntity._

  override def serializers: immutable.Seq[JsonSerializer[_]] = Cmd.serializers ++ Evt.serializers ++ State.serializers

}

class TestEntity(system: ActorSystem)
  extends PersistentEntity {
  import TestEntity._

  override type Command = Cmd
  override type Event = Evt
  override type State = TestEntity.State

  def this(system: ActorSystem, probe: Option[ActorRef]) = {
    this(system)
    this.probe = probe
  }

  var probe: Option[ActorRef] = None

  override def initialState: State = State.empty

  override def behavior: Behavior = {
    case State(Mode.Append, _)  => appending
    case State(Mode.Prepend, _) => prepending
  }

  private val changeMode: Actions = {
    Actions()
      .onCommand[ChangeMode, Evt] {
        case (ChangeMode(mode), ctx, state) => {
          mode match {
            case mode if state.mode == mode => ctx.done
            case Mode.Append                => ctx.thenPersist(InAppendMode)(ctx.reply)
            case Mode.Prepend               => ctx.thenPersist(InPrependMode)(ctx.reply)
          }
        }
      }
  }

  val baseActions: Actions = {
    Actions()
      .onReadOnlyCommand[Get.type, State] {
        case (Get, ctx, state) => ctx.reply(state)
      }
      .onReadOnlyCommand[GetAddress.type, Address] {
        case (GetAddress, ctx, state) => ctx.reply(Cluster.get(system).selfAddress)
      }
      .orElse(changeMode)
  }

  private val appending: Actions =
    baseActions
      .onEvent {
        case (Appended(elem), state) => state.add(elem)
        case (InPrependMode, state)  => state.copy(mode = Mode.Prepend)
      }
      .onCommand[Add, Evt] {
        case (Add(elem, times), ctx, state) =>
          // note that null should trigger NPE, for testing exception
          if (elem == null)
            throw new SimulatedNullpointerException
          if (elem.length == 0) {
            ctx.invalidCommand("element must not be empty");
            ctx.done
          }
          val appended = Appended(elem.toUpperCase)
          if (times == 1)
            ctx.thenPersist(appended)(ctx.reply)
          else
            ctx.thenPersistAll(List.fill(times)(appended): _*)(() => ctx.reply(appended))
      }

  private val prepending: Actions =
    baseActions
      .onEvent {
        case (Prepended(elem), state) => state.add(elem)
        case (InAppendMode, state)    => state.copy(mode = Mode.Append)
      }
      .onCommand[Add, Evt] {
        case (Add(elem, times), ctx, state) =>
          if (elem == null || elem.length == 0) {
            ctx.invalidCommand("element must not be empty");
            ctx.done
          }
          val prepended = Prepended(elem.toLowerCase)
          if (times == 1)
            ctx.thenPersist(prepended)(ctx.reply)
          else
            ctx.thenPersistAll(List.fill(times)(prepended): _*)(() => ctx.reply(prepended))
      }

  override def recoveryCompleted(state: State): State = {
    probe.foreach(_ ! AfterRecovery(state))
    state
  }

}

File: persistence/javadsl/src/test/java/com/lightbend/lagom/javadsl/persistence/TestEntity.java
/*
 * Copyright (C) 2016-2017 Lightbend Inc. <https://www.lightbend.com>
 */
package com.lightbend.lagom.javadsl.persistence;

import akka.Done;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Address;
import akka.cluster.Cluster;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.google.common.collect.ImmutableList;
import com.lightbend.lagom.javadsl.persistence.testkit.SimulatedNullpointerException;
import com.lightbend.lagom.serialization.Jsonable;
import com.lightbend.lagom.javadsl.persistence.PersistentEntity.ReplyType;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class TestEntity extends PersistentEntity<TestEntity.Cmd, TestEntity.Evt, TestEntity.State> {

  public static interface Cmd extends Jsonable {
  }

  public static class Get implements Cmd, ReplyType<State> {

    private static final long serialVersionUID = 1L;

    private static Get instance = new Get();

    @JsonCreator
    public static Get instance() {
      return Get.instance;
    }

    private Get() {
    }
  }

  public static class Add implements Cmd, ReplyType<Evt> {

    private static final long serialVersionUID = 1L;

    private final String element;
    private final int times;

    public static Add of(String element) {
      return new Add(element, 1);
    }

    @JsonCreator
    public Add(String element, int times) {
      this.element = element;
      this.times = times;
    }

    public String getElement() {
      return element;
    }

    public int getTimes() {
      return times;
    }

    @Override
    public int hashCode() {
      final int prime = 31;
      int result = 1;
      result = prime * result + ((element == null) ? 0 : element.hashCode());
      result = prime * result + times;
      return result;
    }

    @Override
    public boolean equals(Object obj) {
      if (this == obj)
        return true;
      if (obj == null)
        return false;
      if (getClass() != obj.getClass())
        return false;
      Add other = (Add) obj;
      if (element == null) {
        if (other.element != null)
          return false;
      } else if (!element.equals(other.element))
        return false;
      if (times != other.times)
        return false;
      return true;
    }

    @Override
    public String toString() {
      return "Add [element=" + element + ", times=" + times + "]";
    }

  }

  public enum Mode {
    PREPEND, APPEND
  }

  public static class ChangeMode implements Cmd, ReplyType<Evt> {

    private static final long serialVersionUID = 1L;

    private final Mode mode;

    @JsonCreator
    public ChangeMode(Mode mode) {
      this.mode = mode;
    }

    public Mode getMode() {
      return mode;
    }

    @Override
    public int hashCode() {
      final int prime = 31;
      int result = 1;
      result = prime * result + ((mode == null) ? 0 : mode.hashCode());
      return result;
    }

    @Override
    public boolean equals(Object obj) {
      if (this == obj)
        return true;
      if (obj == null)
        return false;
      if (getClass() != obj.getClass())
        return false;
      ChangeMode other = (ChangeMode) obj;
      if (mode != other.mode)
        return false;
      return true;
    }

    @Override
    public String toString() {
      return "ChangeMode [mode=" + mode + "]";
    }

  }

  public static class UndefinedCmd implements Cmd, ReplyType<Done> {

    private static final long serialVersionUID = 1L;

    @Override
    public int hashCode() {
      return 0;
    }

    @Override
    public boolean equals(Object obj) {
      if (this == obj)
        return true;
      else
        return (getClass() == obj.getClass());
    }

    @Override
    public String toString() {
      return "UndefinedCmd";
    }
  }

  public static abstract class Evt implements AggregateEvent<Evt>, Jsonable {

    private static final long serialVersionUID = 1L;

    public static final int NUM_SHARDS = 4;

    public static final AggregateEventShards<Evt> AGGREGATE_EVENT_SHARDS = AggregateEventTag.sharded(Evt.class,
        NUM_SHARDS); // second param is optional, defaults to the class name

    @Override
    public AggregateEventShards<Evt> aggregateTag() {
      return AGGREGATE_EVENT_SHARDS;
    }
  }

  public static class Appended extends Evt {

    private static final long serialVersionUID = 1L;

    private final String entityId;
    private final String element;

    @JsonCreator
    public Appended(String entityId, String element) {
      this.entityId = entityId;
      this.element = element;
    }

    public String getEntityId() {
      return entityId;
    }

    public String getElement() {
      return element;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;

      Appended appended = (Appended) o;

      if (!entityId.equals(appended.entityId)) return false;
      return element.equals(appended.element);

    }

    @Override
    public int hashCode() {
      int result = entityId.hashCode();
      result = 31 * result + element.hashCode();
      return result;
    }

    @Override
    public String toString() {
      return "Appended{" +
              "entityId='" + entityId + '\'' +
              ", element='" + element + '\'' +
              '}';
    }
  }

  public static class Prepended extends Evt {

    private static final long serialVersionUID = 1L;

    private final String entityId;
    private final String element;

    @JsonCreator
    public Prepended(String entityId, String element) {
      this.entityId = entityId;
      this.element = element;
    }

    public String getEntityId() {
      return entityId;
    }

    public String getElement() {
      return element;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;

      Prepended prepended = (Prepended) o;

      if (!entityId.equals(prepended.entityId)) return false;
      return element.equals(prepended.element);

    }

    @Override
    public int hashCode() {
      int result = entityId.hashCode();
      result = 31 * result + element.hashCode();
      return result;
    }

    @Override
    public String toString() {
      return "Prepended{" +
              "entityId='" + entityId + '\'' +
              ", element='" + element + '\'' +
              '}';
    }
  }

  public static class InPrependMode extends Evt {

    private static final long serialVersionUID = 1L;

    private final String entityId;

    @JsonCreator
    public InPrependMode(String entityId) {
      this.entityId = entityId;
    }

    public String getEntityId() {
      return entityId;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;

      InPrependMode that = (InPrependMode) o;

      return entityId.equals(that.entityId);

    }

    @Override
    public int hashCode() {
      return entityId.hashCode();
    }

    @Override
    public String toString() {
      return "InPrependMode{" +
              "entityId='" + entityId + '\'' +
              '}';
    }
  }

  public static class InAppendMode extends Evt {

    private static final long serialVersionUID = 1L;

    private final String entityId;

    @JsonCreator
    public InAppendMode(String entityId) {
      this.entityId = entityId;
    }

    public String getEntityId() {
      return entityId;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;

      InAppendMode that = (InAppendMode) o;

      return entityId.equals(that.entityId);

    }

    @Override
    public int hashCode() {
      return entityId.hashCode();
    }

    @Override
    public String toString() {
      return "InAppendMode{" +
              "entityId='" + entityId + '\'' +
              '}';
    }
  }

  public static class State implements Jsonable {

    private static final long serialVersionUID = 1L;

    public static final State EMPTY = new State(Mode.APPEND, ImmutableList.of());

    private final Mode mode;
    private final ImmutableList<String> elements;

    @JsonCreator
    public State(Mode mode, ImmutableList<String> elements) {
      this.mode = mode;
      this.elements = elements;
    }

    public ImmutableList<String> getElements() {
      return elements;
    }

    public State add(String elem) {
      List<String> newElements = new ArrayList<>(elements);
      if (mode == Mode.PREPEND)
        newElements.add(0, elem);
      else
        newElements.add(elem);
      return new State(mode, ImmutableList.copyOf(newElements));
    }

    public State prependMode() {
      return new State(Mode.PREPEND, elements);
    }

    public State appendMode() {
      return new State(Mode.APPEND, elements);
    }

    public Mode getMode() {
      return mode;
    }

    @Override
    public int hashCode() {
      final int prime = 31;
      int result = 1;
      result = prime * result + ((elements == null) ? 0 : elements.hashCode());
      result = prime * result + ((mode == null) ? 0 : mode.hashCode());
      return result;
    }

    @Override
    public boolean equals(Object obj) {
      if (this == obj)
        return true;
      if (obj == null)
        return false;
      if (getClass() != obj.getClass())
        return false;
      State other = (State) obj;
      if (elements == null) {
        if (other.elements != null)
          return false;
      } else if (!elements.equals(other.elements))
        return false;
      if (mode != other.mode)
        return false;
      return true;
    }

    @Override
    public String toString() {
      return "State [mode=" + mode + ", elements=" + elements + "]";
    }

  }

  public static class GetAddress implements Cmd, ReplyType<Address> {

    private static final long serialVersionUID = 1L;

    public static GetAddress instance = new GetAddress();

    private GetAddress() {
    }
  }

  // TestProbe message
  public static class Snapshot {
    public final State state;

    Snapshot(State state) {
      this.state = state;
    }
  }

  //TestProbe message
  public static class AfterRecovery {
    public final State state;

    AfterRecovery(State state) {
      this.state = state;
    }
  }

  private final ActorSystem system;
  private final Optional<ActorRef> probe;

  @Inject
  public TestEntity(ActorSystem system) {
    this.system = system;
    this.probe = Optional.empty();
  }

  public TestEntity(ActorSystem system, ActorRef probe) {
    this.system = system;
    this.probe = Optional.ofNullable(probe);
  }

  @Override
  public Behavior initialBehavior(Optional<State> snapshotState) {

    if (snapshotState.isPresent())
      probe.ifPresent(p -> p.tell(new Snapshot(snapshotState.get()), ActorRef.noSender()));

    BehaviorBuilder b = newBehaviorBuilder(State.EMPTY);

    if (snapshotState.isPresent()) {
      b.setState(snapshotState.get());
    }

    b.setEventHandler(Appended.class, evt -> state().add(evt.getElement()));
    b.setEventHandler(Prepended.class, evt -> state().add(evt.getElement()));

    b.setEventHandlerChangingBehavior(InAppendMode.class, evt -> becomeAppending(behavior()));
    b.setEventHandlerChangingBehavior(InPrependMode.class, evt -> becomePrepending(behavior()));

    b.setReadOnlyCommandHandler(Get.class, (cmd, ctx) -> {
      ctx.reply(state());
    });

    b.setReadOnlyCommandHandler(GetAddress.class, (cmd, ctx) -> {
      ctx.reply(Cluster.get(system).selfAddress());
    });

    b.setCommandHandler(ChangeMode.class,
      (cmd, ctx) -> {
        if (state().getMode() == cmd.getMode()) {
          return ctx.done();
        } else if (cmd.getMode() == Mode.APPEND) {
          return ctx.thenPersist(new InAppendMode(entityId()), evt -> ctx.reply(evt));
        } else if (cmd.getMode() == Mode.PREPEND) {
          return ctx.thenPersist(new InPrependMode(entityId()), evt -> ctx.reply(evt));
        } else {
          throw new IllegalStateException();
        }
      });


    if (b.getState().getMode() == Mode.APPEND) {
      return becomeAppending(b.build());
    } else {
      return becomePrepending(b.build());
    }
  }

  private Behavior becomeAppending(Behavior current) {
    BehaviorBuilder b = current.builder();
    b.setState(current.state().appendMode());
    b.setCommandHandler(Add.class, (cmd, ctx) -> {
      // note that null should trigger NPE, for testing exception
        if (cmd.getElement() == null)
          throw new SimulatedNullpointerException();
      if (cmd.getElement().length() == 0) {
        ctx.invalidCommand("element must not be empty");
        return ctx.done();
      }
      Appended a = new Appended(entityId(), cmd.element.toUpperCase());
      if (cmd.getTimes() == 1)
        return ctx.thenPersist(a, evt -> ctx.reply(evt));
      else
        return ctx.thenPersistAll(fill(a, cmd.getTimes()), () -> ctx.reply(a));
    });
    return b.build();
  }

  private Behavior becomePrepending(Behavior current) {
    BehaviorBuilder b = current.builder();
    b.setState(current.state().prependMode());
    b.setCommandHandler(Add.class, (cmd, ctx) -> {
      if (cmd.getElement() == null || cmd.getElement() == "") {
        ctx.invalidCommand("element must not be empty");
        return ctx.done();
      }
      Prepended a = new Prepended(entityId(), cmd.element.toLowerCase());
      if (cmd.getTimes() == 1)
        return ctx.thenPersist(a, evt -> ctx.reply(evt));
      else
        return ctx.thenPersistAll(fill(a, cmd.getTimes()), () -> ctx.reply(a));
    });
    return b.build();
  }

  private <T> List<T> fill(T evt, int times) {
    List<T> events = new ArrayList<>();
    for (int i = 0; i < times; i++) {
      events.add(evt);
    }
    return events;
  }

  @Override
  public Behavior recoveryCompleted() {
    probe.ifPresent(p -> p.tell(new AfterRecovery(state()), ActorRef.noSender()));
    return behavior();
  }

}
